src/components/Pursuance/views/TaskListView.js

```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import TaskHierarchy from '../../Content/TaskHierarchy/TaskHierarchy'
+
+const TaskListView = ({ match: { params: { pursuanceId } } }) => {
+  return (
+    <TaskHierarchy pursuanceId={pursuanceId} />
+  );
+};
+
+export default TaskListView;
```

Comment:
@Spetastian We'll have a TaskHierarchy and a TaskList.  I'm not sure how we should structure that, but `/tasks` should probably show one of these and `/tasks/hierarchy` or `/tasks/list` should show the other.

For the demo we probably just need `TaskHierarchy`, but we will definitely need both soon.

One thing that may make this easier is: I realized the other day that we can literally use the same exact code to render a hierarchy and a list by saying, "start with the root task(s) and, recursively for each sub-(sub-sub-...)task, render the task then render its sub-tasks indented."

This works if we pass an array of tasks _with no subtasks_ to such a function/method to make the list, and if we pass an array of nested tasks _with subtasks_ to such a function.

Thoughts?

Reply:
Redux! :tada:

Reply:
Let's use Redux, yes... but how? :-) I'm not sure how the tasks should be structured in Redux given that we need to have multiple views of them that preferably load very quickly.

Reply:
That is a good idea. A TaskHierarchy/TaskTree component could might as well render a plain list or with a maximum depth of 2-3 or unlimited, since it will be recursive. I think we should aim at that.

Reply:
My ~latest thinking on the Redux design is at https://github.com/PursuanceProject/pursuance/issues/23 but it's pretty awkward.

Each task should live at just 1 location, and when it is updated then it should just be updated in that one place.  This shouldn't be _too_ hard since obviously JavaScript objects are references, so having references to each task in 2 places in Redux is totally fine; there is nothing to get out of sync, since each task is stored in just 1 place.

Reply:
Each task's subtasks could either show up as references in a `tasks.subtasks` array field, or we could have `task.subtask_gids` or similar where each task's subtasks are referenced by `gid` (which is globally unique per-task and is build from the string `<:pursuanceId>_<:taskId>`).  The latter would be slower due to the additional lookups rather than having task references directly in each `task.subtasks` array, but I'm still thinking through the trade-offs, or other designs.

Reply:
Well, we might have stored the task hierarchy in memory or LocalStorage outside of Redux. Then each time we request a subset of the Heirarchy or a flattened list of some depth a Redux middleware or sorts (could be ReduxThunk or something else) will access that data, aggregate it and dispatch an action do Redux with the correct data. Then all Redux does is update the "task" list for that specific area and the component gets its props and renders.

Reply:
Regarding updating objects via reference. That we should not do in Redux at all. Reducers are pure functions without side effects and should only create a new state, not mutate the current one.

Reply:
Hmm good point, I forgot about the pure function thing.  I'll think more about this...

We don't need to _update_ tasks in place, I'm just thinking about how they should be connected... but oh duh, if we "update" a task in Redux then we're replacing it with a new object, in which case the old references won't point where they need to.

But using each `task.gid` as a(n indirect) "reference" should work a lot better :+1: .